### PR TITLE
Multisig audit fixes

### PR DESCRIPTION
**Pull request by alyn509**
A series of fixes based on the following audit:

https://docs.google.com/document/d/16mn01lNSjmQ36eKa0aH9m84n5YhB-qy7soXUBIiY-vE/edit

### CODE REVIEW

File: contracts/multisig/src/multisig_propose.rs [hunk 1]
```diff
@@ -190,6 +201,12 @@ pub trait MultisigProposeModule: crate::multisig_state::MultisigStateModule {
                     own_shard == other_sc_shard,
                     "All transfer exec must be to the same shard"
                 );
+            } else if let Action::SendTransferExecuteEsdt(call_data) = &action {
```
Comment (ccorcoveanu) [inline contracts/multisig/src/multisig_propose.rs:204]: Can we please refactor the propose_batch function? It's starting to become crowded. We can remove some conditionals into small functions.

File: contracts/multisig/src/multisig_sign.rs [hunk 2]
```diff
@@ -73,6 +92,11 @@ pub trait MultisigSignModule:
         let (caller_id, caller_role) = self.get_caller_id_and_role();
         require!(caller_role.can_sign(), "only board members can un-sign");
 
+        let group_status = self.action_group_status(group_id).get();
+        require!(
+            group_status == ActionStatus::Available,
+            "cannot sign actions of an aborted batch"
```
Comment (ccorcoveanu) [inline contracts/multisig/src/multisig_sign.rs:98]: Wrong error message here.
+ now I think because of this, users cannot discard a batch if it was aborted, thus remove it from state, right?
I can look further but a unit test will be better here

File: contracts/multisig/src/multisig_perform.rs [hunk 3]
```diff
@@ -84,7 +85,7 @@ pub trait MultisigPerformModule:
     /// Returns `true` (`1`) if `getActionValidSignerCount >= getQuorum`.
     #[view(quorumReached)]
     fn quorum_reached(&self, action_id: ActionId) -> bool {
-        let quorum = self.quorum().get();
+        let quorum = self.quorum_for_action(action_id).get();
```
[the remark this answers is not in the document]
Comment (alyn509) [reply contracts/multisig/src/multisig_perform.rs:88]: noted down

File: contracts/multisig/src/multisig_perform.rs [hunk 4]
```diff
@@ -164,6 +192,10 @@ pub trait MultisigPerformModule:
             Action::Nothing => OptionalValue::None,
             Action::AddBoardMember(board_member_address) => {
                 self.change_user_role(action_id, board_member_address, UserRole::BoardMember);
+                require!(
```
[the remark this answers is not in the document]
Comment (alyn509) [reply contracts/multisig/src/multisig_perform.rs:195]: this change will also enforce the change the change of the condition inside the require, since it was initially designed to change the number of board members and to check after if it still meets the requirements

File: contracts/multisig/src/multisig_perform.rs [hunk 5]
```diff
@@ -118,6 +119,21 @@ pub trait MultisigPerformModule:
...
 
+    fn try_perform_action(&self, action_id: ActionId) -> OptionalValue<ManagedAddress> {
+        let (_, caller_role) = self.get_caller_id_and_role();
+        require!(
+            caller_role.can_perform_action(),
+            "only board members and proposers can perform actions"
+        );
+        if self.quorum_reached(action_id) {
+            let group_id = self.group_for_action(action_id).get();
+            require!(group_id == 0, "May not execute this action by itself");
```
[the remark this answers is not in the document]
Comment (alyn509) [reply contracts/multisig/src/multisig_perform.rs:131]: added a custom implementation for it which is also efficient

File: contracts/multisig/src/multisig_perform.rs [hunk 6]
```diff
@@ -148,11 +170,17 @@ pub trait MultisigPerformModule:
     fn perform_action(&self, action_id: ActionId) -> OptionalValue<ManagedAddress> {
         let action = self.action_mapper().get(action_id);
 
+        let group_id = self.group_for_action(action_id).get();
+        let group_status = self.action_group_status(group_id).get();
+        require!(
```
[the remark this answers is not in the document]
Comment (alyn509) [reply contracts/multisig/src/multisig_perform.rs]: I changed this part so that an action has no group it is not checked for group availability. I will make a custom test as well tu make sure it works

File: contracts/multisig/src/multisig_perform.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (alyn509) [reply contracts/multisig/src/multisig_perform.rs:197]: for AddProposer the check needs to remain after change_user_role since at that point we are not sure if we are talking about a new user or a current board member becoming a poposer, so we don't know at that moment if num_board_members will increase or will remain unchanged

File: contracts/multisig/src/multisig.rs [hunk 7]
```diff
@@ -60,12 +58,33 @@ pub trait Multisig:
     /// - the serialized action data
     /// - (number of signers followed by) list of signer addresses.
     #[label("multisig-external-view")]
+    #[allow_multiple_var_args]
```
Comment (dorin-iancu) [inline contracts/multisig/src/multisig.rs:61]: This annotation isn't needed if you use two OptionalValue.

File: contracts/multisig/src/multisig.rs [hunk 8]
```diff
@@ -60,12 +58,33 @@ pub trait Multisig:
...
     #[view(getPendingActionFullInfo)]
-    fn get_pending_action_full_info(&self) -> MultiValueEncoded<ActionFullInfo<Self::Api>> {
+    fn get_pending_action_full_info(
+        &self,
+        count: OptionalValue<usize>,
+        offset_id: OptionalValue<usize>,
```
Comment (dorin-iancu) [inline contracts/multisig/src/multisig.rs:66]: It's usually better to have `opt_` in front of optional args, makes for cleaner code. Then, you wouldn't need to call it `unwrapped_`.

File: contracts/multisig/src/multisig_state.rs [hunk 9]
```diff
@@ -14,9 +23,12 @@ pub trait MultisigStateModule {
     #[storage_mapper("quorum")]
     fn quorum(&self) -> SingleValueMapper<usize>;
 
-    #[storage_mapper("user")]
+    #[storage_mapper("sc_user")]
     fn user_mapper(&self) -> UserMapper;
 
+    #[storage_mapper("quorum_for_action")]
+    fn quorum_for_action(&self, action_id: ActionId) -> SingleValueMapper<usize>;
```
Comment (dorin-iancu) [inline contracts/multisig/src/multisig_state.rs:30]: Why was this needed?

Comment (alyn509) [reply contracts/multisig/src/multisig_state.rs:30]: document - A.3.

File: contracts/multisig/src/multisig.rs (hunk 7 above)
Comment (alyn509) [reply contracts/multisig/src/multisig.rs:61]: `Invalid contract variant multisig-full: Multiple var args found in get_pending_action_full_info. Use #[allow_multiple_var_args] if you want to enable this feature`

the contract build disagrees with you

**Review by dorin-iancu — APPROVED** [review]

File: contracts/multisig/src/multisig_perform.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ccorcoveanu) [reply contracts/multisig/src/multisig_perform.rs:197]: *note to myself - the AddProposer functionality makes no sense - by "Adding a proposer" we are actually "removing the signing right" for the user if he had it previously. This is an unexpected side-effect

**Review by ccorcoveanu — APPROVED** [review]

**Review by ccorcoveanu — CHANGES_REQUESTED** [review]
@alyn509 there are still failing tests in the CI

**Review by dorin-iancu — DISMISSED** [review]

**Review by CostinCarabas — APPROVED** [review]

File: contracts/multisig/src/multisig_perform.rs [hunk 10]
```diff
@@ -256,7 +289,7 @@ pub trait MultisigPerformModule:
             Action::SendAsyncCall(call_data) => {
                 let gas = call_data
                     .opt_gas_limit
-                    .unwrap_or_else(|| self.blockchain().get_gas_left());
+                    .unwrap_or_else(|| self.ensure_and_get_gas_for_transfer_exec());
```
[the remark this answers is not in the document]
Comment (alyn509) [reply contracts/multisig/src/multisig_perform.rs:292]: we will take this change in consider in a future version

File: contracts/multisig/src/multisig_perform.rs [hunk 11]
```diff
@@ -269,6 +302,7 @@ pub trait MultisigPerformModule:
                     .contract_call::<()>(call_data.to, call_data.endpoint_name)
                     .with_egld_transfer(call_data.egld_amount)
                     .with_raw_arguments(call_data.arguments.into())
+                    .with_gas_limit(gas)
```
[the remark this answers is not in the document]
Comment (alyn509) [reply contracts/multisig/src/multisig_perform.rs:305]: we will take this change in consider in a future version

File: contracts/multisig/src/multisig_propose.rs [hunk 12]
```diff
@@ -200,5 +198,41 @@ pub trait MultisigProposeModule: crate::multisig_state::MultisigStateModule {
...
+            "Invalid action"
+        );
+    }
+
+    fn ensure_valid_transfer_action(&self, action: &Action<Self::Api>) {
+        let own_sc_address = self.blockchain().get_sc_address();
+        let own_shard = self.blockchain().get_shard_of_address(&own_sc_address);
+
+        if let Action::SendTransferExecuteEgld(call_data) = &action {
+            let other_sc_shard = self.blockchain().get_shard_of_address(&call_data.to);
```
[the remark this answers is not in the document]
Comment (alyn509) [reply contracts/multisig/src/multisig_propose.rs:217]: we will take this change in consider in a future version

**Review by CostinCarabas — APPROVED** [review]

**Review by ccorcoveanu — APPROVED** [review]
